activesupport/lib/active_support/core_ext/module/delegation.rb

```diff
@@ -199,7 +199,35 @@ def delegate(*methods, to: nil, prefix: nil, allow_nil: nil, private: nil)
 
       # Attribute writer methods only accept one argument. Makes sure []=
       # methods still accept two arguments.
-      definition = /[^\]]=\z/.match?(method) ? "arg" : "..."
+      definition = \
+        if /[^\]]=\z/.match?(method)
+          "arg"
+        else
+          method_object =
+            begin
+              if to.is_a?(Module)
+                to.method(method)
+              elsif receiver == "self.class"
+                method(method)
+              end
+            rescue NameError
+              # Do nothing. Fall back to `"..."`
+            end
+
+          if method_object
+            parameters = method_object.parameters
+
+            if (parameters.map(&:first) & [:opt, :rest, :keyreq, :key, :keyrest]).any?
+              "..."
+            else
+              defn = parameters.filter_map { |type, arg| arg if type == :req }
+              defn << "&block" if parameters.any? { |type, _| type == :block }
```

Comment:
Block should always be last right? So no need to iterate the whole list again.

```suggestion
              defn << "&block" if parameters.last.first == :block
```

Reply:
❗❗ You're a genius!